From ffb6156e5f8531dbc59105d8833eda83947f297b Mon Sep 17 00:00:00 2001
From: Florinel Iordache <florinel.iordache@nxp.com>
Date: Wed, 5 Dec 2018 10:53:05 +0200
Subject: [PATCH 750/767] net/phy: xgkr: Fixed BaseKR algorithm inconsistent
 coefficient updates

This is a corner case but when it occurs it affects the outcome of the BaseKR algorithm and the link training fails. In the case when coefficient are in the state 'NOT UPDATED' the status reported inside the algorithm is 'UPDATED' which is not consistent with the actual operation performed.

Signed-off-by: Florinel Iordache <florinel.iordache@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.12.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/net/phy/fsl_backplane.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/net/phy/fsl_backplane.c b/drivers/net/phy/fsl_backplane.c
index 2a3458324b74..a17fb99e7834 100644
--- a/drivers/net/phy/fsl_backplane.c
+++ b/drivers/net/phy/fsl_backplane.c
@@ -934,13 +934,15 @@ static int inc_dec(struct xgkr_params *xgkr, int field, int request)
 		if (!is_value_allowed((const u32 *)&preq_table, ld_coe[2])) {
 			dev_dbg(&xgkr->phydev->mdio.dev,
 				"preq skipped value: %d\n", ld_coe[2]);
-			return 0;
+			/* NOT UPDATED */
+			return 3;
 		}
 
 		if (!is_value_allowed((const u32 *)&pst1q_table, ld_coe[0])) {
 			dev_dbg(&xgkr->phydev->mdio.dev,
 				"pst1q skipped value: %d\n", ld_coe[0]);
-			return 0;
+			/* NOT UPDATED */
+			return 3;
 		}
 
 		tune_tecr(xgkr);
@@ -953,12 +955,13 @@ static int inc_dec(struct xgkr_params *xgkr, int field, int request)
 			return 2;
 	}
 
+	/* UPDATED */
 	return 0;
 }
 
 static void min_max_updated(struct xgkr_params *xgkr, int field, int new_ld)
 {
-	u32 ld_coe[] = {COE_UPDATED, COE_MIN, COE_MAX};
+	u32 ld_coe[] = {COE_UPDATED, COE_MIN, COE_MAX, COE_NOTUPDATED};
 	u32 mask, val;
 
 	switch (field) {
-- 
2.17.0

